File: Semester3/Advanced Methods of Programming/Toy Language Interpreter C#/Assignement7/Assignement7/Model/CloseStatement.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Assignement7.Model
{
    public class CloseStatement : IStatement
    {
        private Expression expression;

        public CloseStatement(Expression expression)
        {
            this.expression = expression;
        }

        public override string ToString()
        {
            return " closeRFile(" + this.expression.ToString() + ")";
        }

        public ProgramState Execute(ProgramState state)
        {
            int id = this.expression.Evaluate(state.getSymbolTable());
            Tuple<string, TextReader> pair = state.getFileTable().Get(id);

            if (pair == null)
                throw new Exception("FileNotOpened Exception at: " + this.ToString() + "\nThere is no opened file with id = " + id);
            pair.Item2.Close();
            state.getFileTable().Remove(id);
            return state;
        }
    }
}

File: Semester3/Advanced Methods of Programming/ExamenMAP/src/model/statements/file/CloseStatement.java
package model.statements.file;

import model.ProgramState;
import model.exceptions.FileException;
import model.expressions.Expression;
import model.statements.IStatement;
import model.util.MyFilePair;
import myCollections.MyIDictionary;

import java.io.BufferedReader;
import java.io.IOException;

public class CloseStatement implements IStatement {
    private Expression var_name;
    public CloseStatement(Expression var_name) {
        this.var_name = var_name;
    }

    @Override
    public String toString() {
        return " closeRFile(" + var_name.toString() + ")";
    }

    @Override
    public ProgramState execute(ProgramState state) throws Exception {
        MyIDictionary<String, Integer> symbolTable = state.getSymbolTable();
        MyIDictionary<Integer, Integer> heapTable = state.getHeapTable();
        MyIDictionary<Integer, MyFilePair> fileTable = state.getFileTable();
        Integer uniq = var_name.evaluate(symbolTable, heapTable);

        MyFilePair pair = fileTable.get(uniq);
        if(pair == null)
            throw new FileException("file does not exist!!\n");

        BufferedReader b = pair.getBufferedReader();

        try{
            b.close();
        } catch (IOException e) {
            throw new FileException("cannot close this file!\n");
        }

        fileTable.remove(uniq);

        return null;
    }
}
